backends/ebpf/ebpfParser.cpp

```diff
@@ -73,6 +74,49 @@ StateTranslationVisitor::compileLookahead(const IR::Expression* destination) {
     builder->blockEnd(true);
 }
 
+void
+StateTranslationVisitor::compileAdvance(const P4::ExternMethod* extMethod) {
+    auto argExpr = extMethod->expr->arguments->at(0)->expression;
+    cstring argStr;
+    if (auto cnst = argExpr->to<IR::Constant>()) {
+        argStr = cstring::to_cstring(cnst->asUnsigned());
+    } else {
+        argStr = argExpr->toString();
+    }
+    cstring offsetStr = Util::printf_format("BYTES(%s + %s)",
+                                            state->parser->program->offsetVar, argStr);
+    builder->target->emitTraceMessage(builder, "Parser (advance): check pkt_len=%%d < "
+                                               "last_read_byte=%%d", 2,
+                                      state->parser->program->lengthVar.c_str(),
+                                      offsetStr.c_str());
+    builder->emitIndent();
+    builder->appendFormat("if (%s < %s + BYTES(%s + ",
+                          state->parser->program->packetEndVar.c_str(),
+                          state->parser->program->packetStartVar.c_str(),
+                          state->parser->program->offsetVar.c_str());
+    visit(argExpr);
```

Comment:
This is the right way to do it.
Moreover, you may want to save the result of the evaluation in a temporary so you don't evaluate the expression multiple times. There should be no side-effects, so I think the result will always be correct, but perhaps inefficient.

Reply:
it is not about the compile-time cost, it is about the runtime cost.
You will evaluate the expression twice.

Reply:
Do you mean eBPF program performance? By the way we can implement advance with only one `visit`

Reply:
yes, ebpf performance,
The number of 'visit' calls does not matter.

Reply:
Yes, you're right. I had in mind constant cases before. Assuming we will fix non-constant advance it will affect a performance